### PR TITLE
Flask Application Insights Sample

### PR DESCRIPTION
**Pull request by daveta**
Creates a Flask + Application Insights sample based on Core-Bot

In particular:
- Flask Middleware to perform post-body caching. 
- Fixes Application Insights Telemetry Processor.  Add explicit return values for filtering, and add better documentation.
- Removed some dead code in Django Middleware that was converting POST body to Json object when it didn't need to.

### CODE REVIEW

File: libraries/botbuilder-dialogs/botbuilder/dialogs/waterfall_dialog.py [hunk 1]
```diff
@@ -1,150 +1,151 @@
...
+                    List)
+from .dialog_reason import DialogReason
+from .dialog import Dialog
+from .dialog_turn_result import DialogTurnResult
+from .dialog_context import DialogContext
+from .dialog_instance import DialogInstance
+from .waterfall_step_context import WaterfallStepContext
+from botbuilder.core import TurnContext
+from botbuilder.schema import ActivityTypes
+from typing import Coroutine, List
```
[the remark this answers is not in the document]
Comment (daveta) [reply libraries/botbuilder-dialogs/botbuilder/dialogs/waterfall_dialog.py]: Fixed!  Thanks

File: samples/python-flask/21.corebot-app-insights/bots/dialog_and_welcome_bot.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import json
+import os.path
+
+from typing import List
+from botbuilder.core import CardFactory
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/bots/dialog_and_welcome_bot.py]: Fixed - thanks!

File: samples/python-flask/21.corebot-app-insights/config.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class DefaultConfig(object):
+    PORT = 3978
+    APP_ID = ""
+    APP_PASSWORD = ""
+    LUIS_APP_ID = ""
+    LUIS_API_KEY = ""
+    LUIS_API_HOST_NAME = ""
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/config.py]: Added examples, thanks!!

File: samples/python-flask/21.corebot-app-insights/config.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class DefaultConfig(object):
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/config.py]: Fixed, thanks!

File: samples/python-flask/21.corebot-app-insights/dialogs/cancel_and_help_dialog.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+from botbuilder.core import BotTelemetryClient, NullTelemetryClient
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/dialogs/cancel_and_help_dialog.py]: Fixed, thanks!

File: samples/python-flask/21.corebot-app-insights/dialogs/main_dialog.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+from botbuilder.core import BotTelemetryClient, NullTelemetryClient
+from botbuilder.dialogs import ComponentDialog, DialogSet, DialogTurnStatus, WaterfallDialog, WaterfallStepContext, DialogTurnResult
+from botbuilder.dialogs.prompts import TextPrompt, ConfirmPrompt, PromptOptions
+from botbuilder.core import MessageFactory
+from .booking_dialog import BookingDialog
+from booking_details import BookingDetails
+from helpers.luis_helper import LuisHelper
+from datatypes_date_time.timex import Timex
+
+class MainDialog(ComponentDialog):
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/dialogs/main_dialog.py]: Fixed, thanks!

File: samples/python-flask/21.corebot-app-insights/helpers/activity_helper.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+from datetime import datetime
+from botbuilder.schema import Activity, ActivityTypes, ChannelAccount, ConversationAccount
+
+def create_activity_reply(activity: Activity, text: str = None, locale: str = None):
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/helpers/activity_helper.py]: Fixed, thanks!

File: libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/django/__init__.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
+"""Djanjo Application Insights package."""
```
Comment (stevengum) [inline libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/django/__init__.py]: Typo here, should be `"Django"`

File: libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/flask/flask_telemetry_middleware.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+# Map of thread id => POST body text
+_REQUEST_BODIES = {}
+
+def retrieve_flask_body():
+    """ retrieve_flask_body
+    Retrieve the POST body text from temporary cache.
+    The POST body corresponds with the thread id and should resides in
+    cache just for lifetime of request.
+
+    TODO: Add cleanup job to kill orphans
```
Comment (stevengum) [inline libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/flask/flask_telemetry_middleware.py]: Could we perhaps clarify this `TODO`? Is this completed or should we file a separate tracking bug for the future?

File: samples/python-flask/21.corebot-app-insights/dialogs/main_dialog.py [hunk 10]
```diff
@@ -37,38 +39,57 @@ def __init__(self, configuration: dict, dialog_id: str = None, telemetry_client:
...
+                PromptOptions(prompt=MessageFactory.text("What can I help you with today?"))) # pylint: disable=bad-continuation
 
 
     async def act_step(self, step_context: WaterfallStepContext) -> DialogTurnResult:
-        # Call LUIS and gather any potential booking details. (Note the TurnContext has the response to the prompt.)
-        booking_details = await LuisHelper.excecute_luis_query(self._configuration, step_context.context) if step_context.result is not None else BookingDetails()
+        """Use language understanding to gather details about booking."""
+        # Call LUIS and gather any potential booking details. (Note the TurnContext
+        # has the response to the prompt.)
+        print(f'LUIS HEKOPE: {self.telemetry_client._instrumentation_key}')
```
Comment (stevengum) [inline samples/python-flask/21.corebot-app-insights/dialogs/main_dialog.py]: I think I'm missing some context here, what's a `LUIS HEKOPE`? :)

File: libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/django/bot_telemetry_middleware.py [hunk 11]
```diff
@@ -1,21 +1,21 @@
...
+_REQUEST_BODIES = {}
 
 def retrieve_bot_body():
     """ retrieve_bot_body
     Retrieve the POST body text from temporary cache.
-    The POST body corresponds with the thread id and should resides in 
+    The POST body corresponds with the thread id and should resides in
     cache just for lifetime of request.
 
     TODO: Add cleanup job to kill orphans
```
Comment (stevengum) [inline libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/django/bot_telemetry_middleware.py]: Could we perhaps clarify this TODO? Is this completed or should we file a separate tracking bug for the future?

File: samples/python-flask/21.corebot-app-insights/dialogs/main_dialog.py (hunk 10 above)
Comment (daveta) [reply samples/python-flask/21.corebot-app-insights/dialogs/main_dialog.py]: Heh, good find!

File: libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/flask/flask_telemetry_middleware.py (hunk 9 above)
Comment (daveta) [reply libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/flask/flask_telemetry_middleware.py]: I removed it comment in both cases.  To hit this, the bots pretty much broken.

File: libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/django/bot_telemetry_middleware.py (hunk 11 above)
Comment (daveta) [reply libraries/botbuilder-applicationinsights/botbuilder/applicationinsights/django/bot_telemetry_middleware.py]: Same as above..

**Pull request merged**